packages/flutter_test/lib/src/binding.dart

```diff
@@ -1476,20 +1481,19 @@ class LiveTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
   }) {
     switch (source) {
       case TestBindingEventSource.test:
-        if (!renderView._pointers.containsKey(event.pointer)) {
-          assert(event.down || event is PointerAddedEvent);
-          if (event.down) {
-            renderView._pointers[event.pointer] = _LiveTestPointerRecord(
-              event.pointer,
-              event.position,
-            );
-          }
-        } else {
+        if (renderView._pointers.containsKey(event.pointer)) {
           renderView._pointers[event.pointer].position = event.position;
           if (!event.down)
             renderView._pointers[event.pointer].decay = _kPointerDecay;
+          _handleViewNeedsPaint();
+        } else if (event.down) {
+          assert(event is PointerDownEvent);
+          renderView._pointers[event.pointer] = _LiveTestPointerRecord(
+            event.pointer,
+            event.position,
+          );
+          _handleViewNeedsPaint();
```

Comment:
By having `_handleViewNeedsPaint` inside the `if else if` branch, it seems that `_handleViewNeedsPaint` won't be called for hover events now but it was called before this change? I wonder if that's intended, and whether hover should trigger some painting.

Reply:
It's intended, because 
1. the original version, if I neglect the problematic `assert` sentence, basically doesn't draw hover event;
2. If we are to draw hover event and scroll event, I would think a better practice is to have a separate list/map for that and keep the original one just for down event, because it's more meaningful to plot them in a different way.

Reply:
To paint hover event, would probably better to be a TODO, if we decide to paint it.

Reply:
To clarify, `paint` method for this class is only painting `_pointers` data. So if all events are down-move-up, than it always needs paint. But for other events, it originally should violate the assert, and nothing will be added to `_pointers`, so it's a fake (literally, not in the test sense) `_handleViewNeedsPaint`.

Reply:
Discussed with @CareF offline. This `_handleViewNeedsPaint` only triggers repaint of the `_pointers` positions which do not include non-down (e.g., hover) pointers. So we don't have to call it if it's not down or already tracked in `_pointers`. The repaint of the test app should be driven by `super.dispatchEvent` and that should be the place where the test app responds to hover repaint.